storage/src/main/java/tech/pegasys/artemis/storage/CombinedChainDataClient.java

```diff
@@ -0,0 +1,129 @@
+import tech.pegasys.artemis.datastructures.util.BeaconStateUtil;
+
+public class CombinedChainDataClient {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private static final CompletableFuture<Optional<BeaconBlock>> BLOCK_NOT_AVAILABLE =
+      completedFuture(Optional.empty());
+  private final ChainStorageClient recentChainData;
+  private final HistoricalChainData historicalChainData;
+
+  public CombinedChainDataClient(
+      final ChainStorageClient recentChainData, final HistoricalChainData historicalChainData) {
+    this.recentChainData = recentChainData;
+    this.historicalChainData = historicalChainData;
+  }
+
+  /**
+   * Returns the block proposed for the requested slot on the chain identified by <code>
+   * headBlockRoot</code>. If the slot was empty, no block is returned.
+   *
+   * @param slot the slot to get the block for
+   * @param headBlockRoot the block root of the head of the chain
+   * @return the block at the requested slot or empty if the slot was empty
+   */
+  public CompletableFuture<Optional<BeaconBlock>> getBlockAtSlotExact(
+      final UnsignedLong slot, final Bytes32 headBlockRoot) {
+    return getBlockInEffectAtSlot(slot, headBlockRoot)
+        .thenApply(maybeBlock -> maybeBlock.filter(block -> block.getSlot().equals(slot)));
+  }
+
+  /**
+   * Returns the block which was proposed in or most recently before the requested slot on the chain
+   * specified by <code>headBlockRoot</code>. If the slot was empty, the block at the last filled
+   * slot is returned.
+   *
+   * @param slot the slot to get the effective block for
+   * @param headBlockRoot the block root of the head of the chain
+   * @return the block at slot or the closest previous slot if empty
+   */
+  public CompletableFuture<Optional<BeaconBlock>> getBlockInEffectAtSlot(
+      final UnsignedLong slot, final Bytes32 headBlockRoot) {
+    final Store store = recentChainData.getStore();
+    if (store == null) {
+      LOG.trace("No block at slot {} because the store is not set", slot);
+      return BLOCK_NOT_AVAILABLE;
+    }
+
+    final BeaconState headState = store.getBlockState(headBlockRoot);
+    if (headState == null) {
+      LOG.trace("No block at slot {} because head block root {} is unknown", slot, headBlockRoot);
+      return BLOCK_NOT_AVAILABLE;
+    }
+    if (headState.getSlot().compareTo(slot) < 0) {
+      LOG.trace(
+          "No block at slot {} because it is after the referenced head state slot {}",
+          slot,
+          headState.getSlot());
+      return BLOCK_NOT_AVAILABLE;
+    }
+    if (headState.getSlot().equals(slot)) {
+      LOG.trace("Block root at slot {} is the specified head block root", slot);
+      return completedFuture(Optional.ofNullable(store.getBlock(headBlockRoot)));
+    }
+    if (isFinalized(slot)) {
+      LOG.trace("Block at slot {} is in a finalized epoch. Retrieving from historical data", slot);
+      return historicalChainData.getFinalizedBlockAtSlot(slot);
```

Comment:
I'm a bit confused with the lines 95 and 98 here. What's the difference between historicalChainData.getFinalizedBlockAtSlot and getBlockAtSlotFormHistoricalBlockRoot?

Reply:
and if a slot is not finalized, why would we use `getBlockAtSlotFormHistoricalBlockRoots` to retrieve the block associated with it?

Reply:
I think the idea is that eventually all finalized blocks are stored in `historicalChainData`, while all the non-finalized blocks are kept in memory in the `Store`.  So, if it's finalized, we go to the database (`historicalChainData`) to pull the value.  If it's not, we look for it in the historical blocks kept in the state associated with the head block.